File: src/agent_bench/runner.py
"""Task runner for executing benchmarks."""

import logging
import shutil
import time
from dataclasses import dataclass
from pathlib import Path
from typing import List

from git import Repo

from .agents import Agent, AgentType, create_agent
from .error import AgentError, GitError, VerificationError
from .evaluator import BenchmarkResult, SuiteResults, Verifier
from .task import Task, TaskLoader

logger = logging.getLogger(__name__)


@dataclass
class RunnerConfig:
    """Configuration for the task runner."""

    tasks_dir: Path
    results_dir: Path
    workspace_dir: Path
    max_iterations: int = 1

    @classmethod
    def default(cls) -> "RunnerConfig":
        """Create a default configuration."""
        return cls(
            tasks_dir=Path("tasks"),
            results_dir=Path("results"),
            workspace_dir=Path("/tmp/agent-bench"),
            max_iterations=1,
        )


class TaskRunner:
    """Task runner for executing benchmarks."""

    def __init__(self, config: RunnerConfig):
        self.config = config

    @classmethod
    def with_defaults(cls) -> "TaskRunner":
        """Create a task runner with default configuration."""
        return cls(RunnerConfig.default())

    async def run_task(self, task_id: str, agent_type: AgentType) -> BenchmarkResult:
        """Run a single task with the specified agent."""
        loader = TaskLoader(self.config.tasks_dir)
        task = loader.load_by_id(task_id)
        agent = create_agent(agent_type)

        return await self._execute_task(task, agent)

    async def run_all(self, agent_type: AgentType) -> SuiteResults:
        """Run all tasks with the specified agent."""
        loader = TaskLoader(self.config.tasks_dir)
        tasks = loader.load_all()
        agent = create_agent(agent_type)

        results = []
        for task in tasks:
            print(f"Running task: {task.id} - {task.title}")
            result = await self._execute_task(task, agent)
            status = "PASS" if result.success else "FAIL"
            print(f"  Result: {status} (score: {result.score}, duration: {result.duration_secs:.2f}s)")
            results.append(result)

        suite = SuiteResults.from_results(agent.name(), results)
        path = suite.save(self.config.results_dir)
        print(f"\nSuite results saved to: {path}")

        return suite

    async def _execute_task(self, task: Task, agent: Agent) -> BenchmarkResult:
        """Execute a single task."""
        start = time.time()

        # Prepare workspace
        try:
            workspace = self._prepare_workspace(task)
        except Exception as e:
            duration = time.time() - start
            return BenchmarkResult.create_failure(
                task_id=task.id,
                agent=agent.name(),
                iterations=0,
                tokens_used=None,
                duration_secs=duration,
                error=f"Failed to prepare workspace: {e}",
                agent_version=None,
                model_name=None,
            )

        # Execute agent
        try:
            agent_result = await agent.execute(task, workspace)
        except Exception as e:
            duration = time.time() - start
            return BenchmarkResult.create_failure(
                task_id=task.id,
                agent=agent.name(),
                iterations=0,
                tokens_used=None,
                duration_secs=duration,
                error=f"Agent execution failed: {e}",
                agent_version=None,
                model_name=None,
            )

        # Run verification
        try:
            verification = await Verifier.verify(task, workspace)
        except Exception as e:
            duration = time.time() - start
            result = BenchmarkResult.create_failure(
                task_id=task.id,
                agent=agent.name(),
                iterations=agent_result.iterations,
                tokens_used=agent_result.tokens_used,
                duration_secs=duration,
                error=f"Verification failed: {e}",
                agent_version=agent_result.agent_version,
                model_name=agent_result.model_name,
            )
            result.with_agent_output(agent_result.output)
            path = result.save(self.config.results_dir)
            print(f"Result saved to: {path}")
            return result

        duration = time.time() - start

        # Create result
        if verification.passed:
            result = BenchmarkResult.create_success(
                task_id=task.id,
                agent=agent.name(),
                iterations=agent_result.iterations,
                tokens_used=agent_result.tokens_used,
                duration_secs=duration,
                agent_version=agent_result.agent_version,
                model_name=agent_result.model_name,
            )
        else:
            result = BenchmarkResult.create_failure(
                task_id=task.id,
                agent=agent.name(),
                iterations=agent_result.iterations,
                tokens_used=agent_result.tokens_used,
                duration_secs=duration,
                error="Verification tests failed",
                agent_version=agent_result.agent_version,
                model_name=agent_result.model_name,
            )

        result.with_agent_output(agent_result.output)
        result.with_verification_output(
            f"Exit code: {verification.exit_code}\n\n"
            f"STDOUT:\n{verification.stdout}\n\n"
            f"STDERR:\n{verification.stderr}"
        )

        # Save individual result
        path = result.save(self.config.results_dir)
        print(f"Result saved to: {path}")

        return result

    def _prepare_workspace(self, task: Task) -> Path:
        """Prepare a workspace for task execution."""
        workspace = self.config.workspace_dir / task.id

        # Clean up existing workspace if it exists
        if workspace.exists():
            shutil.rmtree(workspace)

        # Clone the repository only if not "none"
        if task.source.repository != "none" and task.source.repository:
            self._clone_repo(task.source.repository, task.source.commit, workspace)
        else:
            # Create empty workspace directory for tasks that don't need a repository
            workspace.mkdir(parents=True, exist_ok=True)

        return workspace

    def _clone_repo(self, repo_url: str, commit: str, workspace: Path) -> None:
        """Clone a repository to the workspace."""
        try:
            # Clone the repository
            repo = Repo.clone_from(repo_url, str(workspace))

            # If commit is "main", "master", or "HEAD", stay on default branch
            if commit in {"main", "master", "HEAD"}:
                return

            # Try to checkout as a commit hash or branch
            try:
                repo.git.checkout(commit)
            except Exception as e:
                raise GitError(f"Failed to checkout '{commit}': {e}")

        except Exception as e:
            raise GitError(f"Failed to clone repository: {e}")

    def list_tasks(self) -> List[Task]:
        """List all available tasks."""
        loader = TaskLoader(self.config.tasks_dir)
        return loader.load_all()

File: src/core/runner.ts
/**
 * Task runner for executing benchmarks.
 */

import { TaskLoader } from './loader.js';
import { WorkspaceManager } from './workspace.js';
import { Task } from './task.js';
import type { Agent } from '../agents/types.js';
import { Verifier } from '../evaluator/verifier.js';
import type { BenchmarkResult } from '../evaluator/results.js';
import {
  createSuccess,
  createFailure,
  withAgentOutput,
  withVerificationOutput,
  saveResult,
  createSuiteResults,
  saveSuiteResults,
} from '../evaluator/results.js';
import type { RunnerConfig } from './config.js';
import { logger } from '../utils/logger.js';


/**
 * Task runner for executing benchmarks.
 */
export class TaskRunner {
  private config: RunnerConfig;
  private loader: TaskLoader;
  private workspace: WorkspaceManager;

  constructor(config: RunnerConfig) {
    this.config = config;
    this.loader = new TaskLoader(config.tasksDir);
    this.workspace = new WorkspaceManager(config.workspaceDir);
  }

  /**
   * Run a single task with the specified agent.
   */
  async runTask(
    taskId: string,
    agent: Agent,
    skipVerify: boolean = false,
    runId?: string
  ): Promise<BenchmarkResult> {
    const task = await this.loader.loadById(taskId);
    return await this.executeTask(task, agent, skipVerify, runId);
  }

  /**
   * Run a single task against multiple models in parallel.
   */
  async runTaskParallel(
    taskId: string,
    runs: Array<{ agent: Agent; runId: string }>,
    skipVerify: boolean = false
  ): Promise<BenchmarkResult[]> {
    const task = await this.loader.loadById(taskId);

    return await Promise.all(
      runs.map(({ agent, runId }) => this.executeTask(task, agent, skipVerify, runId))
    );
  }

  /**
   * Run all tasks.
   */
  async runAll(agent: Agent, skipVerify: boolean = false, runId?: string): Promise<void> {
    const tasks = await this.loader.loadAll();

    const runLabel = runId ? `${agent.name()} (${runId})` : agent.name();
    logger.info(`Running ${tasks.length} tasks with agent: ${runLabel}`);

    const results: BenchmarkResult[] = [];

    for (const task of tasks) {
      logger.taskHeader(task.id, task.title);

      const result = await this.executeTask(task, agent, skipVerify, runId);
      results.push(result);

      logger.taskResult(
        result.success,
        result.score,
        result.iterations,
        result.duration_secs,
        result.tokens_used || undefined
      );
    }

    // Save suite results
    const suite = createSuiteResults(runLabel, results);
    const suitePath = await saveSuiteResults(suite, this.config.resultsDir);

    logger.suiteSummary(
      suite.total_tasks,
      suite.passed,
      suite.failed,
      suite.pass_rate,
      suite.total_duration_secs
    );

    logger.success(`Suite results saved to: ${suitePath}`);
  }

  /**
   * Run tasks by category.
   */
  async runCategory(
    category: string,
    agent: Agent,
    skipVerify: boolean = false,
    runId?: string
  ): Promise<void> {
    const tasks = await this.loader.filterByCategory(category);

    if (tasks.length === 0) {
      logger.warn(`No tasks found for category: ${category}`);
      return;
    }

    const runLabel = runId ? `${agent.name()} (${runId})` : agent.name();
    logger.info(`Running ${tasks.length} tasks in category "${category}" with agent: ${runLabel}`);

    const results: BenchmarkResult[] = [];

    for (const task of tasks) {
      logger.taskHeader(task.id, task.title);

      const result = await this.executeTask(task, agent, skipVerify, runId);
      results.push(result);

      logger.taskResult(
        result.success,
        result.score,
        result.iterations,
        result.duration_secs,
        result.tokens_used || undefined
      );
    }

    // Save suite results
    const suite = createSuiteResults(runLabel, results);
    const suitePath = await saveSuiteResults(suite, this.config.resultsDir);

    logger.suiteSummary(
      suite.total_tasks,
      suite.passed,
      suite.failed,
      suite.pass_rate,
      suite.total_duration_secs
    );

    logger.success(`Suite results saved to: ${suitePath}`);
  }

  /**
   * Execute a single task.
   */
  private async executeTask(
    task: Task,
    agent: Agent,
    skipVerify: boolean,
    runId?: string
  ): Promise<BenchmarkResult> {
    const startTime = Date.now();

    // Prepare workspace
    logger.info('Preparing workspace...');
    let workspacePath: string;
    try {
      workspacePath = await this.workspace.prepare(task, runId);
      logger.success(`Workspace ready: ${workspacePath}`);
    } catch (error) {
      const duration = (Date.now() - startTime) / 1000;
      const result = createFailure(
        task.id,
        agent.name(),
        0,
        null,
        duration,
        `Failed to prepare workspace: ${error}`,
        null,
        null
      );
      await saveResult(result, this.config.resultsDir);
      return result;
    }

    try {
      // Determine the directory the agent should operate in.
      // When source.run_path is set the agent is scoped to that subdirectory
      // so it cannot accidentally touch files belonging to other tasks.
      const agentPath = this.workspace.getAgentPath(task, runId);
      if (agentPath !== workspacePath) {
        logger.debug(`Agent scoped to run_path: ${agentPath}`);
      }

      // Execute agent
      logger.info('Executing agent...');
      let agentResult;
      try {
        agentResult = await agent.execute(task, agentPath);
        logger.success(`Agent execution completed: ${agentResult.iterations} iterations`);
      } catch (error) {
        const duration = (Date.now() - startTime) / 1000;
        const result = createFailure(
          task.id,
          agent.name(),
          0,
          null,
          duration,
          `Agent execution failed: ${error}`,
          null,
          null
        );
        await saveResult(result, this.config.resultsDir);
        return result;
      }

      // Run verification (unless skipped)
      let result: BenchmarkResult;
      if (skipVerify) {
        logger.warn('Skipping verification');
        result = createSuccess(
          task.id,
          agent.name(),
          agentResult.iterations,
          agentResult.tokensUsed,
          agentResult.durationSecs,
          agentResult.agentVersion,
          agentResult.modelName,
          agentResult.inputTokens,
          agentResult.outputTokens
        );
      } else {
        logger.info('Running verification...');
        try {
          const verification = await Verifier.verify(task, workspacePath);

          if (verification.passed) {
            logger.success('Verification passed');
            result = createSuccess(
              task.id,
              agent.name(),
              agentResult.iterations,
              agentResult.tokensUsed,
              agentResult.durationSecs,
              agentResult.agentVersion,
              agentResult.modelName,
              agentResult.inputTokens,
              agentResult.outputTokens
            );
          } else {
            logger.error(`Verification failed with exit code: ${verification.exitCode}`);
            result = createFailure(
              task.id,
              agent.name(),
              agentResult.iterations,
              agentResult.tokensUsed,
              agentResult.durationSecs,
              'Verification tests failed',
              agentResult.agentVersion,
              agentResult.modelName,
              agentResult.inputTokens,
              agentResult.outputTokens
            );
          }

          // Add verification output
          result = withVerificationOutput(
            result,
            `Exit code: ${verification.exitCode}\n\nSTDOUT:\n${verification.stdout}\n\nSTDERR:\n${verification.stderr}`
          );
        } catch (error) {
          logger.error(`Verification error: ${error}`);
          result = createFailure(
            task.id,
            agent.name(),
            agentResult.iterations,
            agentResult.tokensUsed,
            agentResult.durationSecs,
            `Verification error: ${error}`,
            agentResult.agentVersion,
            agentResult.modelName,
            agentResult.inputTokens,
            agentResult.outputTokens
          );
        }
      }

      // Add agent output
      result = withAgentOutput(result, agentResult.output);

      // Save result
      const resultPath = await saveResult(result, this.config.resultsDir);
      logger.debug(`Result saved to: ${resultPath}`);

      // Only clean up workspace on success; preserve it on failure for debugging
      if (result.success) {
        await this.workspace.cleanup(task, runId);
      } else {
        logger.warn(`Task failed — workspace preserved for debugging: ${agentPath}`);
      }

      return result;
    } catch (error) {
      // Unexpected error — preserve workspace for debugging
      logger.warn(`Unexpected error — workspace preserved for debugging: ${workspacePath}`);
      throw error;
    }
  }

  /**
   * List all available tasks.
   */
  async listTasks(): Promise<Task[]> {
    return await this.loader.loadAll();
  }
}
